http-body-util/src/throttle.rs

```diff
@@ -0,0 +1,163 @@
+use bytes::Buf;
+use http::HeaderMap;
+use http_body::{Body, SizeHint};
+use pin_project_lite::pin_project;
+use std::{
+    convert::{TryFrom, TryInto},
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::{sleep, Instant, Sleep};
+
+#[derive(Debug)]
+enum State {
+    Waiting(Pin<Box<Sleep>>, Instant),
+    Ready(Instant),
+    Init,
+}
+
+pin_project! {
+    /// A throttled body.
+    #[derive(Debug)]
+    #[cfg_attr(docsrs, doc(cfg(feature = "tokio")))]
+    pub struct Throttle<B> {
+        #[pin]
+        inner: B,
+        state: State,
+        cursor: f64,
+        byte_rate: f64,
+    }
+}
+
+impl<B> Throttle<B> {
+    /// Create a new `Throttle`.
+    ///
+    /// # Panic
+    ///
+    /// Will panic if milliseconds in `duration` is larger than `u32::MAX`.
+    pub fn new(body: B, duration: Duration, bytes: u32) -> Self {
+        let bytes = f64::from(bytes);
+        let duration = f64::from(u32::try_from(duration.as_millis()).expect("duration too large"));
+
+        let byte_rate = bytes / duration;
+
+        Self {
+            inner: body,
+            state: State::Init,
+            cursor: 0.0,
+            byte_rate,
+        }
+    }
+}
+
+impl<B: Body> Body for Throttle<B> {
+    type Data = B::Data;
+    type Error = B::Error;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        let mut this = self.project();
+
+        loop {
+            match this.state {
+                State::Waiting(sleep, time) => match sleep.as_mut().poll(cx) {
+                    Poll::Ready(()) => {
+                        let byte_rate = *this.byte_rate;
+                        let mut elapsed = to_f64(time.elapsed().as_millis());
+
+                        if elapsed > 2000.0 {
+                            elapsed = 2000.0;
+                        }
+
+                        *this.cursor += elapsed * byte_rate;
+                        *this.state = State::Ready(Instant::now());
+                    }
+                    Poll::Pending => return Poll::Pending,
+                },
+                State::Ready(time) => match this.inner.as_mut().poll_data(cx) {
```

Comment:
Of note, if we get a really large, single chunk, then no real throttling of data is done. Instead, this does throttling of chunk pulls, which may be pretty coarse in practice, as the max buffer size in hyper for HTTP/1 is 408 kiB. With a `Full` inner body , `Throttle` would send everything as a single chunk without any throttling.

To implement throttling regardless of chunk size, you may need to hold the underlying data as buffer to enable re-chunking the data on the way through. potentially avoiding floating-point rate calculations.

One way would be: call `poll_data()`, split off up to quota and send (or all if below quota), if quota reached, then save away remaining bytes and halt sending until the next time horizon. On reaching next time horizon, send quota out of remaining bytes. If bytes are exhausted, poll data again and repeat. There’s some decision to be made here between number and size of chunks and the ability to track the requested throttle rate. With large time buckets, you can keep chunks relatively large, but will end up with highly-variable instantaneous throughput. With small time buckets, you may have a smoother throughput profile, but have more overhead in the number of chunks being sent downstream.

Thoughts?

Reply:
> To implement throttling regardless of chunk size, you may need to hold the underlying data as buffer to enable re-chunking the data on the way through. potentially avoiding floating-point rate calculations.

I think this buffering cost should be optional to users. Maybe a `Buffer` body utility can be added.

> Instead, this does throttling of chunk pulls, which may be pretty coarse in practice, as the max buffer size in hyper for HTTP/1 is 408 kiB.

Can't really get around that except documenting this setting and having users set it.

Reply:
> I think this buffering cost should be optional to users. Maybe a `Buffer` body utility can be added.

In practice, as the chunks are `Bytes`, the split operation is cheap and doesn’t actually require allocating any distinct memory.